### PR TITLE
Fault Tolerance Refinements

### LINKED ISSUE #13: Provide separate hook for handling insufficient funds
InsufficientFunds exceptions incurred when calling `sendRawTransaction` either during broadcast or tx retry, should be handled separately. This is a well known error case that should be handled by a callback defined by the user (eg. `on_insufficient_funds`).

Not much else the atx machine can do.

The user can decide what to do eg. pause the machine see #10 .

### PR DESCRIPTION
**Pull request by derekpierre**
- [X] Don't requeue txs after failure since order can be of significance. Txs remain in their position in the queue in spite of failure, unless the user takes more drastic action
- [X] `on_broadcast_failure` and `on_fault` callbacks are now mandatory when queueing a transaction. It ensures that the user has thought about what to do during a failure scenario.
- [X] A reverted tx (receipt with status 0) is not actually a fault, it is really just a `FinalizedTx` that was unsuccessful; don't treat it as a fault anymore. As a result the `on_finalized` callback is now mandatory when queueing a transaction since the user should handle the case where the finalized tx was reverted / unsuccessful.
- [X] Be more efficient with RPC calls as they relate to checking transaction confirmations.
- [X] Allow the speedup strategy to specify the minimum time between speedup increases - allows for more control over overspending for a tx.
- [X] Provide separate mandatory callback for the case of insufficient funds - fixes #13.
- [x] Update repos diagram to reflect updates.

### CODE REVIEW

**Review by vzotova — APPROVED** [review]
LGTM (mostly high level understanding only)

File: atxm/machine.py [hunk 1]
```diff
@@ -352,33 +331,39 @@ def __strategize(self) -> None:
...
-        self.log.info(f"[retry] transaction #{pending_tx.id} has been re-broadcasted")
+        self.log.info(
+            f"[retry] transaction #{pending_tx.id} has been re-broadcasted with updated params"
+        )
         if pending_tx.on_broadcast:
             fire_hook(hook=pending_tx.on_broadcast, tx=pending_tx)
 
     def __handle_retry_failure(self, attempted_tx: PendingTx, e: Exception):
-        log.warn(
+        self.log.warn(
```
Comment (manumonti) [inline atxm/machine.py:360]: should we do the same on lines 225 and 235?

File: atxm/machine.py [hunk 2]
```diff
@@ -520,3 +512,12 @@ def queue_transaction(
             self._wake()
 
         return tx
+
+    def remove_queued_transaction(self, tx: FutureTx):
+        """
+        Removes a queued transaction; useful when an exisitng queued transaction has broadcast
```
Comment (manumonti) [inline atxm/machine.py:518]:
```suggestion
        Removes a queued transaction; useful when an existing queued transaction has broadcast
```

**Review by manumonti — CHANGES_REQUESTED** [review]
The code overall looks great 🙌

Although I think I caught some typos.

File: diagram.png
[the remark this answers is not in the document]
Comment (derekpierre) [reply diagram.png:1]: ✔️

File: tests/test_machine.py [hunk 3]
```diff
@@ -314,6 +361,133 @@ def test_broadcast(
...
+@pytest_twisted.inlineCallbacks
+def test_broadcast_insufficient_funds(
+    clock,
+    w3,
+    machine,
+    legacy_transaction,
+    account,
+    mocker,
+    mock_wake_sleep,
+):
```
[the remark this answers is not in the document]
Comment (derekpierre) [reply tests/test_machine.py]: ✔️

File: tests/test_machine.py [hunk 4]
```diff
@@ -314,6 +361,133 @@ def test_broadcast(
...
+@pytest_twisted.inlineCallbacks
+def test_broadcast_insufficient_funds_eth_tester(
+    clock,
+    w3,
+    machine,
+    eip1559_transaction,
+    account,
+    mocker,
+    mock_wake_sleep,
+):
```
[the remark this answers is not in the document]
Comment (derekpierre) [reply tests/test_machine.py]: ✔️

File: tests/test_machine.py [hunk 5]
```diff
@@ -888,6 +1100,154 @@ def test_use_strategies_that_dont_make_updates(
...
+    ethereum_tester,
+    w3,
+    machine,
+    state_observer,
+    clock,
+    eip1559_transaction,
+    account,
+    mocker,
+    mock_wake_sleep,
+):
```
[the remark this answers is not in the document]
Comment (derekpierre) [reply tests/test_machine.py]: ✔️ Some can be removed, but `mock_wake_sleep` is necessary.

File: tests/test_machine.py [hunk 6]
```diff
@@ -888,6 +1100,154 @@ def test_use_strategies_that_dont_make_updates(
...
```
[the remark this answers is not in the document]
Comment (derekpierre) [reply tests/test_machine.py]: ✔️ Same as above

File: tests/test_faults.py [hunk 7]
```diff
@@ -99,7 +81,14 @@ def test_strategy_fault(
...
-    w3, machine, clock, eip1559_transaction, account, interval, mock_wake_sleep, mocker
+    w3,
+    machine,
+    clock,
+    eip1559_transaction,
+    account,
+    interval,
+    mock_wake_sleep,
+    mocker,
 ):
```
[the remark this answers is not in the document]
Comment (derekpierre) [reply tests/test_faults.py]: ✔️

File: atxm/machine.py (hunk 1 above)
Comment (derekpierre) [reply atxm/machine.py:360]: ✔️

File: atxm/machine.py (hunk 2 above)
Comment (derekpierre) [reply atxm/machine.py]: ✔️

File: atxm/machine.py [hunk 8]
```diff
@@ -249,35 +246,28 @@ def __handle_active_transaction(self) -> None:
         The 4 possible outcomes for the pending ("active") transaction in one cycle:
```
[the remark this answers is not in the document]
Comment (derekpierre) [reply atxm/machine.py]: ✔️

**Review by piotr-roslaniec — APPROVED** [review]

**Pull request merged**